### PR TITLE
Add attribute output_dtype to QuantizeLinear

### PR DESCRIPTION
**Pull request by galagam**
The purpose of this change is to allow setting the quantized type without providing the zero-point tensor for symmetric quantization.
This reduces model size, most importantly for block quantization where the zero-point tensor dimensions are large, and reduces backend runtime.

This implements issue #5943

### CODE REVIEW

**Comment by galagam** [comment]
@gramalingam following up on our discussion in the Operators SIG meeting yesterday, here are the changes for #5943.
If you can do a quick review, hopefully we'll be able to get this into v1.16.

File: onnx/defs/quantization/defs.cc [hunk 1]
```diff
@@ -97,7 +104,9 @@ ONNX_OPERATOR_SET_SCHEMA(
             "The type of the input `y_zero_point` and the output `y`.")
         .SetDoc(QuantizeLinear_ver21_doc)
         .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
-          if (ctx.hasInput(2)) {
+          if (ctx.getAttribute("output_dtype") != nullptr) {
+            propagateElemTypeFromAttributeToOutput(ctx, "output_dtype", 0);
```
Comment (gramalingam) [inline onnx/defs/quantization/defs.cc:108]: nit: would be useful to check, if zeropoint is also specified, that it has the same type.

Comment (gramalingam) [reply onnx/defs/quantization/defs.cc:108]: I guess output_dtype is also required to be in the set of allowed T2 types.

Comment (galagam) [reply onnx/defs/quantization/defs.cc]: @gramalingam 

> nit: would be useful to check, if zeropoint is also specified, that it has the same type.

I've added a check that zeropoint and output_dtype match if both are specified. Also added type inference tests to check all cases (only zero-point, only output_dtype, both, both mismatch).

> I guess output_dtype is also required to be in the set of allowed T2 types.

True. I was surprised I couldn't find something similar in any other operator. This should be fixed as part of `onnx/defs/shape_inference.h::propagateElemTypeFromAttributeToOutput`, since this affects other operators (i.e. Cast).
I don't have a quick fix for this, but I've filed #5960 to follow up on that. This PR also includes a couple of skipped tests that should be unskipped once #5960 is fixed.

File: onnx/defs/quantization/defs.cc [hunk 2]
```diff
@@ -97,8 +104,21 @@ ONNX_OPERATOR_SET_SCHEMA(
...
         .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
-          if (ctx.hasInput(2)) {
+          auto const zp_type = ctx.hasInput(2) ? ctx.getInputType(2) : nullptr;
+          auto const output_dtype = getAttribute(ctx, "output_dtype", TensorProto::UNDEFINED);
+          if (zp_type != nullptr) {
+            const auto zp_elem_type = getTensorElementType(*zp_type);
+            if (output_dtype != TensorProto::UNDEFINED && output_dtype != zp_elem_type) {
+              fail_type_inference(
+                  "output_dtype ",
+                  ctx.getAttribute("output_dtype"),
```
Comment (gramalingam) [inline onnx/defs/quantization/defs.cc:114]: nit: why not `output_dtype` ? Also `TensorProto_DataType_Name(output_dtype)` will print the symbolic name of the datatype. (Same for `zp_elem_type` below.)

**Review by gramalingam — APPROVED** [review]
LGTM, thanks for the quick PR, greatly appreciate it!

File: onnx/defs/quantization/defs.cc (hunk 2 above)
Comment (gramalingam) [reply onnx/defs/quantization/defs.cc]: Ok, it looks like the CI doesn't like the use of the int type in an enum context. May need an explicit cast from the int type to the enum type.
